src/main/msp/msp_serial.c

```diff
@@ -131,35 +246,116 @@ static uint8_t mspSerialChecksumBuf(uint8_t checksum, const uint8_t *data, int l
 {
+    // We are allowed to send out the response if
+    //  a) TX buffer is completely empty (we are talking to well-behaving party that follows request-response scheduling;
+    //     this allows us to transmit jumbo frames bigger than TX buffer (serialWriteBuf will block, but for jumbo frames we don't care)
+    //  b) Response fits into TX buffer
+    const int totalFrameLength = hdrLen + dataLen + crcLen;
+    if (!isSerialTransmitBufferEmpty(msp->port) && ((int)serialTxBytesFree(msp->port) < totalFrameLength))
+        return 0;
+
+    // Transmit frame
     serialBeginWrite(msp->port);
-    const int len = sbufBytesRemaining(&packet->buf);
-    const int mspLen = len < JUMBO_FRAME_SIZE_LIMIT ? len : JUMBO_FRAME_SIZE_LIMIT;
-    uint8_t hdr[8] = {
-        '$',
-        'M',
-        packet->result == MSP_RESULT_ERROR ? '!' : packet->direction == MSP_DIRECTION_REPLY ? '>' : '<',
-        mspLen,
-        packet->cmd
-    };
-    int hdrLen = 5;
-#define CHECKSUM_STARTPOS 3  // checksum starts from mspLen field
-    if (len >= JUMBO_FRAME_SIZE_LIMIT) {
-        hdrLen += 2;
-        hdr[5] = len & 0xff;
-        hdr[6] = (len >> 8) & 0xff;
-    }
     serialWriteBuf(msp->port, hdr, hdrLen);
-    uint8_t checksum = mspSerialChecksumBuf(0, hdr + CHECKSUM_STARTPOS, hdrLen - CHECKSUM_STARTPOS);
-    if (len > 0) {
-        serialWriteBuf(msp->port, sbufPtr(&packet->buf), len);
-        checksum = mspSerialChecksumBuf(checksum, sbufPtr(&packet->buf), len);
-    }
-    serialWriteBuf(msp->port, &checksum, 1);
+    serialWriteBuf(msp->port, data, dataLen);
+    serialWriteBuf(msp->port, crc, crcLen);
     serialEndWrite(msp->port);
-    return sizeof(hdr) + len + 1; // header, data, and checksum
+
+    return totalFrameLength;
+}
+
+static int mspSerialEncode(mspPort_t *msp, mspPacket_t *packet, mspVersion_e mspVersion)
+{
+    static const uint8_t mspMagic[MSP_VERSION_COUNT] = { 'M', 'M', 'X' };
+    const int dataLen = sbufBytesRemaining(&packet->buf);
+    uint8_t hdrBuf[16] = { '$', mspMagic[mspVersion], packet->result == MSP_RESULT_ERROR ? '!' : '>'};
+    uint8_t crcBuf[2];
+    int hdrLen = 3;
+    int crcLen = 0;
+
+    #define V1_CHECKSUM_STARTPOS 3
+    if (mspVersion == MSP_V1) {
+        mspHeaderV1_t * hdrV1 = (mspHeaderV1_t *)&hdrBuf[hdrLen];
+        hdrLen += sizeof(mspHeaderV1_t);
+        hdrV1->cmd = packet->cmd;
+
+        // Add JUMBO-frame header if necessary
+        if (dataLen >= JUMBO_FRAME_SIZE_LIMIT) {
+            mspHeaderJUMBO_t * hdrJUMBO = (mspHeaderJUMBO_t *)&hdrBuf[hdrLen];
+            hdrLen += sizeof(mspHeaderJUMBO_t);
+
+            hdrV1->size = JUMBO_FRAME_SIZE_LIMIT;
+            hdrJUMBO->size = dataLen;
+        }
+        else {
+            hdrV1->size = dataLen;
+        }
+
+        // Pre-calculate CRC
+        crcBuf[crcLen] = mspSerialChecksumBuf(0, hdrBuf + V1_CHECKSUM_STARTPOS, hdrLen - V1_CHECKSUM_STARTPOS);
```

Comment:
compute CRC in local variable, then append it to `crcBuf`

Reply:
Please explain why? Would the generated code be different?

Reply:
It will explicitly document checksum type/size and will allow consistent migration to larger checksums. Also checksum endianness would be easier to handle (not important here)

Reply:
Good point